Sprint 8 solution time and duration

Just-Roman opened the pull request:
Sprint8:
1. class CSVFormatter:  method toString, add task.getDuration()  task.getStartTime().
2. class Task: add Duration duration, LocalDateTime startTime; add method getDuration(), getStartTime(), getEndTime().
3.class Epic: add LocalDateTime endTime; add method setDurationAndLocalDateTime().
4. class FileBackedTaskManagerTest: add method clearManager(), testException().
5. add class InMemoryHistoryManagerTest.
6. class InMemoryTaskManager: add Set<Task> sortedEntity; add method sorted(Task o1, Task o2); add method synchronizeSortedEntity(); getPrioritizedTasks(); add method checkPeriodCrossing(LocalDateTime startTime, LocalDateTime endTime; changed method calculateStatusEpic(List<Subtask> subtasks); 
7. class InMemoryTaskManagerTest: add method checkTreeset(); 
8.  add abstract class TaskManagerTes ;

Commit:
Sprint8:
1. class CSVFormatter:  method toString, add task.getDuration()  task.getStartTime().
2. class Task: add Duration duration, LocalDateTime startTime; add method getDuration(), getStartTime(), getEndTime().
3.class Epic: add LocalDateTime endTime; add method setDurationAndLocalDateTime().
4. class FileBackedTaskManagerTest: add method clearManager(), testException().
5. add class InMemoryHistoryManagerTest.
6. class InMemoryTaskManager: add Set<Task> sortedEntity; add method sorted(Task o1, Task o2); add method synchronizeSortedEntity(); getPrioritizedTasks(); add method checkPeriodCrossing(LocalDateTime startTime, LocalDateTime endTime; changed method calculateStatusEpic(List<Subtask> subtasks);
7. class InMemoryTaskManagerTest: add method checkTreeset();
8.  add abstract class TaskManagerTes;

Commit:
Sprint8:
1. class CSVFormatter:  method toString, add task.getDuration()  task.getStartTime().
2. class Task: add Duration duration, LocalDateTime startTime; add method getDuration(), getStartTime(), getEndTime().
3.class Epic: add LocalDateTime endTime; add method setDurationAndLocalDateTime().
4. class FileBackedTaskManagerTest: add method clearManager(), testException().
5. add class InMemoryHistoryManagerTest.
6. class InMemoryTaskManager: add Set<Task> sortedEntity; add method sorted(Task o1, Task o2); add method synchronizeSortedEntity(); getPrioritizedTasks(); add method checkPeriodCrossing(LocalDateTime startTime, LocalDateTime endTime; changed method calculateStatusEpic(List<Subtask> subtasks);
7. class InMemoryTaskManagerTest: add method checkTreeset();
8.  add abstract class TaskManagerTes ;

Commit:
Sprint8:
1. class CSVFormatter:  method toString, add task.getDuration()  task.getStartTime().
2. class Task: add Duration duration, LocalDateTime startTime; add method getDuration(), getStartTime(), getEndTime().
3.class Epic: add LocalDateTime endTime; add method setDurationAndLocalDateTime().
4. class FileBackedTaskManagerTest: add method clearManager(), testException().
5. add class InMemoryHistoryManagerTest.
6. class InMemoryTaskManager: add Set<Task> sortedEntity; add method sorted(Task o1, Task o2); add method synchronizeSortedEntity(); getPrioritizedTasks(); add method checkPeriodCrossing(LocalDateTime startTime, LocalDateTime endTime; changed method calculateStatusEpic(List<Subtask> subtasks);
7. class InMemoryTaskManagerTest: add method checkTreeset();
8.  add abstract class TaskManagerTes ;

slavrikk commented on src/ru/practicum/task_tracker/CSVFormatter.java:
```diff
@@ -5,35 +5,39 @@
-                .append(task.getStatus()).append(",")
-                .append(task.getEpicId())
-                .toString();
+        return task.getId() + "," +
```
Корректней оставить `StringBuilder`

slavrikk commented on src/ru/practicum/task_tracker/manager/FileBackedTaskManager.java:
```diff
@@ -20,12 +20,10 @@ public class FileBackedTaskManager extends InMemoryTaskManager {
     private static File file;
```
Лучше сделать переменную не статическую

Just-Roman replied:
Если сделаю НЕ статическую, то из метода loadFromFile(Path path), не смогу создать и вернуть экземпляр класса, FileBackedTaskManager.  Как поступить?

slavrikk commented on src/ru/practicum/task_tracker/manager/FileBackedTaskManager.java:
```diff
@@ -20,12 +20,10 @@ public class FileBackedTaskManager extends InMemoryTaskManager {
     private static File file;
 
     FileBackedTaskManager(File file) {
-        this.file = file;
```
Тут оставить как было

slavrikk commented on src/ru/practicum/task_tracker/manager/FileBackedTaskManager.java:
```diff
@@ -20,12 +20,10 @@ public class FileBackedTaskManager extends InMemoryTaskManager {
+        FileBackedTaskManager.file = file;
     }
 
     public static void loadFromFile(Path path) {
```
Тут лучше возвращать `FileBackedTaskManager`

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryTaskManager.java:
```diff
@@ -44,27 +74,41 @@ public Task getByTaskId(Integer taskId) {
-    public Task createTask(Task task) {
+    public Task createTask(Task task) throws Exception {
+        if (checkPeriodCrossing(task.getStartTime(), task.getEndTime())) {
+            throw new Exception("Период задачи пересекается с текущими");
```
Лучше создать кастомную ошибку, которая будет наследоваться от `RuntimeException()`, тогда все `throws` в сигнатуре метода можно будет удалить

slavrikk replied:
Замечание для всех методов, где есть проверка на пересечения

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryTaskManager.java:
```diff
@@ -6,19 +6,49 @@
     protected static Map<Integer, Subtask> subtasks = new HashMap<>();
     protected static int nextId;
+    protected Set<Task> sortedEntity = new TreeSet<>(this::sorted);
+    HistoryManager historyManager = Managers.getDefaultHistory();
```
Так же можно добавить модификатор `protected`

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryTaskManager.java:
```diff
@@ -6,19 +6,49 @@
+        return sortedEntity;
+    }
+
+    public boolean checkPeriodCrossing(LocalDateTime startTime, LocalDateTime endTime) {
```
Лучше в метод передать `Task` и из этого объекта получать `startTime` и `endTime`

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryTaskManager.java:
```diff
@@ -6,19 +6,49 @@
+    public boolean checkPeriodCrossing(LocalDateTime startTime, LocalDateTime endTime) {
+        return sortedEntity.stream()
+                .anyMatch(e ->
+                        !(e.getStartTime().isBefore(startTime) && e.getEndTime().isBefore(startTime)
```
Достаточно одного такого условия:
`!(end1.isBefore(start2) || start1.isAfter(end2))`
https://www.baeldung.com/java-check-two-date-ranges-overlap

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryTaskManager.java:
```diff
@@ -44,27 +74,41 @@ public Task getByTaskId(Integer taskId) {
         int id = task.getId();
+
         tasks.put(id, task);
+        synchronizeSortedEntity();
```
Лучше сразу делать вставку в коллекцию `sortedEntity`, без очистки и полного добавления

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryTaskManager.java:
```diff
@@ -214,10 +269,11 @@ public void deleteAllTasks() {
     }
 
     private int getNextId() {
-        return nextId++;
```
Чем `++` не устроил?)

slavrikk commented on src/ru/practicum/task_tracker/manager/TaskManager.java:
```diff
@@ -5,18 +5,21 @@
     Task getByTaskId(Integer taskId);
 
-    Task createTask(Task task);
+    Task createTask(Task task) throws Exception;
```
Все `throws` лишние (появнил выше)

slavrikk commented on src/ru/practicum/task_tracker/task/Epic.java:
```diff
@@ -24,6 +32,31 @@ public void setSubtasks(Subtask subtask) {
+                .max(LocalDateTime::compareTo)
+                .orElse(null);
+        assert endDateTime != null;
+        return endDateTime.plus(duration);
```
`endTime` уже подсчитан в `Task`, он равен `startTime.plus(duration)`

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryTaskManager.java:
```diff
@@ -6,19 +6,49 @@
 
-    HistoryManager historyManager = Managers.getDefaultHistory();
-
     protected static Map<Integer, Task> tasks = new HashMap<>();
```
Поля должны быть не статические, чтобы у каждого экземпляра `InMemoryTaskManager` были свои собственные коллекции тасок

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryTaskManager.java:
```diff
@@ -6,19 +6,49 @@
     protected static Map<Integer, Task> tasks = new HashMap<>();
     protected static Map<Integer, Epic> epics = new HashMap<>();
     protected static Map<Integer, Subtask> subtasks = new HashMap<>();
     protected static int nextId;
```
Так же с переменной `nextId`, для каждого экземпляра `InMemoryTaskManager` она уникальна

slavrikk commented on test/ru/practicum/task_tracker/manager/InMemoryTaskManagerTest.java:
```diff
@@ -1,110 +1,153 @@
 
     @Test
-    void checkEqualsCreatedTask() {
+    void checkTreeset() throws Exception {
```
В этом тесте хорошо бы проверить, что таски в коллекции `getPrioritizedTasks()`, находятся в правильном порядке сортировки

Commit:
Sprint8:
1. class FileBackedTaskManagerTest: update method getEndTime()
2. class FileBackedTaskManager: update method FileBackedTaskManager loadFromFile(Path path); deleted signature throws.
3. InMemoryTaskManager: modifier static removed; deleted method synchronizeSortedEntity(); deleted signature throws.
4. InMemoryTaskManagerTest: update method checkTreeSet(); deleted signature throws.
5. class Main: deleted signature throws.
6. ManagerSaveException: add constructor ManagerSaveException(String msg).
7. class ManagersTest: deleted signature throws.
8. interface TaskManager: deleted signature throws.
9. class TaskManagerTest: deleted signature throws.
